# XGBoost with carbon, silicon, and nitrgen isotopes

In [33]:
# Imports
import pandas as pd

from numpy import loadtxt
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from xgboost import XGBClassifier

## Read the CSV

In [34]:
# load data
#dataset = loadtxt('presolargrains_C_only.csv', delimiter=",")
# Read csv file
C_Si_N = pd.read_csv('presolargrains_C_Si_N.csv')
C_Si_N.head()

Type  carbon_12_13  nitrogen_14_15  silicon_29_28  silicon_30_28
0    X         74.30           207.4         -685.0         -520.0
1    X       1581.00           116.0         -684.0         -490.0
2    X        234.75           187.2         -683.0         -501.0
3    X        455.00           140.0         -662.0         -770.0
4    X        823.00            44.0         -658.0         -234.0

## Train Test Split

In [36]:
# split data into X and y
#X = carbon_nitrogen['carbon_12_13', 'nitrogen_14_15']
X = C_Si_N.loc[::1,'carbon_12_13':'silicon_30_28']
y = C_Si_N['Type'] #.values.reshape(-1, 1)
print(X)

      carbon_12_13  nitrogen_14_15  silicon_29_28  silicon_30_28
0            74.30          207.40        -685.00        -520.00
1          1581.00          116.00        -684.00        -490.00
2           234.75          187.20        -683.00        -501.00
3           455.00          140.00        -662.00        -770.00
4           823.00           44.00        -658.00        -234.00
5           140.00           97.00        -653.00        -446.00
6           199.30           68.40        -610.00        -360.00
7           136.70          116.30        -606.00        -195.00
8           145.16          128.50        -601.00        -528.00
9           223.00          102.00        -600.00        -459.00
10          169.60          154.00        -598.00        -523.00
11         1693.89           63.20        -588.00        -605.00
12         2807.00           83.00        -578.00        -468.00
13         9455.10           28.00        -569.00        -647.00
14          216.59       

In [37]:
# Split data into train and test sets
seed = 1
test_size = 0.33
#X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=test_size, random_state=seed)
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1, stratify=y)

## Fit model

In [38]:
# fit model
model = XGBClassifier()
model.fit(X_train, y_train)

XGBClassifier(base_score=0.5, booster=None, colsample_bylevel=1,
       colsample_bynode=1, colsample_bytree=1, gamma=0, gpu_id=-1,
       importance_type='gain', interaction_constraints=None,
       learning_rate=0.300000012, max_delta_step=0, max_depth=6,
       min_child_weight=1, missing=nan, monotone_constraints=None,
       n_estimators=100, n_jobs=0, num_parallel_tree=1,
       objective='multi:softprob', random_state=0, reg_alpha=0,
       reg_lambda=1, scale_pos_weight=None, subsample=1, tree_method=None,
       validate_parameters=False, verbosity=None)

## Print model

In [39]:
# Print model
print(model)

XGBClassifier(base_score=0.5, booster=None, colsample_bylevel=1,
       colsample_bynode=1, colsample_bytree=1, gamma=0, gpu_id=-1,
       importance_type='gain', interaction_constraints=None,
       learning_rate=0.300000012, max_delta_step=0, max_depth=6,
       min_child_weight=1, missing=nan, monotone_constraints=None,
       n_estimators=100, n_jobs=0, num_parallel_tree=1,
       objective='multi:softprob', random_state=0, reg_alpha=0,
       reg_lambda=1, scale_pos_weight=None, subsample=1, tree_method=None,
       validate_parameters=False, verbosity=None)


## Compare predictions to actual

In [40]:
# Make predictions for test data
y_pred = model.predict(X_test)
# Create and print dataframe with predicted and actual types
pd.DataFrame({"Prediction": y_pred, "Actual": y_test}).reset_index(drop=True)

Prediction Actual
0            X      X
1            X      X
2            M      M
3           AB      Y
4            M      M
5            M      M
6           AB     AB
7            M      M
8           AB     AB
9           AB     AB
10           M      M
11           M      M
12           M      M
13           M      M
14           X      X
15           M      M
16           M      M
17           X      X
18           X      X
19           M      M
20           M      M
21           X      X
22           Z      Z
23           Y      Y
24           M      M
25          AB     AB
26           M      M
27          AB     AB
28           M      M
29           M      M
..         ...    ...
296          M      M
297          X      X
298          M      M
299          M      M
300          M      M
301         AB     AB
302         AB     AB
303          M      M
304          Y      Y
305          M     AB
306          M      M
307          M      M
308          M      M
309          Y      X
310         AB     AB
311          M      M
312          M      M
313          X      X
314          M      M
315          Y      Y
316          M      M
317          M      M
318          M      M
319          X      X
320          M      M
321          X      X
322         AB     AB
323          X      X
324          M      M
325          M      M

[326 rows x 2 columns]

## Report Accuracy

In [41]:
# Evaluate predictions
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy: %.2f%%" % (accuracy * 100.0))

Accuracy: 96.32%


In [ ]:
## Save the model

In [ ]:
# Save model to file
filename = 'XGBoost_C_Si_N.sav'
joblib.dump(clf, filename)

## Summary: 

### XGBoost with carbon, silicon, and nitrogen isotopes: 96.3%